vm/src/intern.rs

```diff
@@ -0,0 +1,121 @@
+use crate::{
+    builtins::{PyStr, PyTypeRef},
+    common::lock::PyRwLock,
+    convert::ToPyObject,
+    PyRef, PyRefExact,
+};
+use std::ops::Deref;
+
+#[derive(Debug)]
+pub struct StringPool {
+    inner: PyRwLock<std::collections::HashSet<CachedPyStrRef, ahash::RandomState>>,
+}
+
+impl Default for StringPool {
+    fn default() -> Self {
+        Self {
+            inner: PyRwLock::new(Default::default()),
+        }
+    }
+}
+
+impl Clone for StringPool {
+    fn clone(&self) -> Self {
+        Self {
+            inner: PyRwLock::new(self.inner.read().clone()),
+        }
+    }
+}
+
+impl StringPool {
+    #[inline]
+    pub unsafe fn intern<S: Internable>(&self, s: S, typ: PyTypeRef) -> PyRefExact<PyStr> {
+        if let Some(found) = self.inner.read().get(s.as_str()) {
+            return found.clone().inner;
+        }
+        let cache = CachedPyStrRef {
+            inner: s.into_pyref(typ),
+        };
+        self.inner.write().insert(cache.clone());
+        cache.inner
+    }
+}
+
+#[derive(Debug)]
+pub struct CachedPyStrRef {
+    inner: PyRefExact<PyStr>,
+}
+
+impl Clone for CachedPyStrRef {
+    fn clone(&self) -> Self {
+        Self {
+            inner: unsafe { PyRefExact::new_unchecked(self.inner.clone()) },
```

Comment:
Would probably be good to impl Clone for PyRefExact and then you can just derive